### ISSUE TITLE
Reset input_mask during Generative.

### ISSUE THREAD

**Issue by liuqiangict** [lucidrains/reformer-pytorch#74]
Hi Author, 

I believe the `input_mask `need to be reset during [generate](https://github.com/lucidrains/reformer-pytorch/blob/319a28d418d902c712885cec0b1a77c1877a901c/reformer_pytorch/generative_tools.py#L51), since the output sequence concat one more again and again?

**Comment by lucidrains** [comment]
Oh! You are right! You must be working with batches of variable length sequences. I'll have to think about this one, since I believe the sequence need to be left padded for the generation and masking shifting to be efficient

**Comment by liuqiangict** [comment]
For left padded, I think you can just update the concat at [`x = torch.cat((x, mem, keys), dim=1)`](https://github.com/lucidrains/reformer-pytorch/blob/319a28d418d902c712885cec0b1a77c1877a901c/reformer_pytorch/reformer_pytorch.py#L470) and [` mask = torch.cat((i_mask, m_mask, c_mask), dim=1)`](https://github.com/lucidrains/reformer-pytorch/blob/319a28d418d902c712885cec0b1a77c1877a901c/reformer_pytorch/reformer_pytorch.py#L490). But when I tried update that, still got incorrect generate output. 

The incorrect output (most is 0) caused by the autopadd with[ `F.pad(tensor, (*pad_offset, 0, remainder), value=0)`](https://github.com/lucidrains/reformer-pytorch/blob/319a28d418d902c712885cec0b1a77c1877a901c/reformer_pytorch/autopadder.py#L13), seems `input_mask` or `input_attn_mask` still not working. 

Do you have any ideas how to solve it?

**Comment by lucidrains** [comment]
@liuqiangict I'll have to think about this and get back to you. For now, I think you can still pass in a single sequence, and it should work. It's really the edge case of batched sequences with variable padding that is the problem. The other thing to note is that for sequences shorter than the bucket size, you could default to full attention using the `full_attn_thres` keyword

**Comment by liuqiangict** [comment]
Thanks.
Actually, I am working with the single sequence. 
The problem is, when generate, input only have 1 id, assume it is `start_id`. Due to the `bucket_size` limitation, will `AutoPad` another 127 0s  to `[start_id, 0, 0, ..., 0]`. Then generated output will most like 0. Seems `input_mask` didn't do anything. 

Maybe I misunderstand something about the code, have send one mail to you gmail box, hope can get some more help from you, thanks.

**Comment by lucidrains** [comment]
@liuqiangict ahh i see! please try setting your `full_attn_thres = 128` (you can do this without having to retrain). what it will do is default to full attention for sequence lengths less than 128.

**Comment by liuqiangict** [comment]
Thanks for your quick and nice reply. 

It doesn't work since in generation, the `keys` length mostly large than 128, as [code ](https://github.com/lucidrains/reformer-pytorch/blob/319a28d418d902c712885cec0b1a77c1877a901c/reformer_pytorch/reformer_pytorch.py#L465-L467) showd, `kv_len `have small chance smaller than 128, still not fix the issue. 

The root cause is why `input_mask `not working as well. maybe I have write some wrong code there.

**Comment by lucidrains** [comment]
@liuqiangict Ok, I'll release a makeshift solution that should work for single sequences, still right padded, later today, so that the generate function shifts the input mask appropriately. I'll get a permanent left padded fix out probably tomorrow

**Comment by lucidrains** [comment]
@liuqiangict thank you for identifying this problem!

**Comment by liuqiangict** [comment]
> @liuqiangict Ok, I'll release a makeshift solution that should work for single sequences, still right padded, later today, so that the generate function shifts the input mask appropriately. I'll get a permanent left padded fix out probably tomorrow

**Comment by lucidrains** [comment]
@liuqiangict https://github.com/lucidrains/reformer-pytorch/commit/a5c4b326e70632312f83e6b1f703502fff9e5194 this should solve the autopadding issue, let me know if that works

I will work on having the `input_mask` auto shifted during generate in the next hour, assuming batched initial start tokens of the same length

**Comment by lucidrains** [comment]
@liuqiangict https://github.com/lucidrains/reformer-pytorch/commit/352214dc4d2fb13c6018706ec226e60145a7c857 this will take care of padding the input mask for you when you are generating samples, assuming both your sequence and mask is *left padded*!

**Comment by liuqiangict** [comment]
Thanks Phil.

I guess this [line ](https://github.com/lucidrains/reformer-pytorch/blob/352214dc4d2fb13c6018706ec226e60145a7c857/reformer_pytorch/generative_tools.py#L56)still incorrect. Should use `input_mask = torch.full_like(out, True, dtype=torch.bool, device=out.device)` instead. 

For example, at the very beginning, only given `start_id`, shape of `input_mask` is `[batch_size, ids]`, which is `[1, 1]`. While with `out` generated more and more ids, `input_mask` shape didn't update, still  `[1, 1]`.

**Comment by lucidrains** [comment]
@liuqiangict I believe it is updated here https://github.com/lucidrains/reformer-pytorch/blob/352214dc4d2fb13c6018706ec226e60145a7c857/reformer_pytorch/generative_tools.py#L64 that line is to keep the running window under the `max_seq_len`

**Comment by liuqiangict** [comment]
Oh, I see, missed that line, you are correct.
